main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1155 @@
+      for {
+        terms <- Validation.traverse(terms0)(visitExpression(_, kenv, root))
+      } yield KindedAst.Predicate.Head.Atom(pred, den, terms, Type.freshVar(Kind.Star), loc)
+  }
+
+  /**
+    * Performs kinding on the given body predicate under the given kind environment.
+    */
+  private def visitBodyPredicate(pred: ResolvedAst.Predicate.Body, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Predicate.Body, KindError] = pred match {
+    case ResolvedAst.Predicate.Body.Atom(pred, den, polarity, terms0, loc) =>
+      for {
+        terms <- Validation.traverse(terms0)(visitPattern(_, kenv, root))
+      } yield KindedAst.Predicate.Body.Atom(pred, den, polarity, terms, Type.freshVar(Kind.Star), loc)
+    case ResolvedAst.Predicate.Body.Guard(exp0, loc) =>
+      for {
+        exp <- visitExpression(exp0, kenv, root)
+      } yield KindedAst.Predicate.Body.Guard(exp, loc)
+  }
+
+  /**
+    * Performs kinding on the given type variable under the given kind environment, with `kindMatch` expected from context.
+    */
+  private def visitTypeVar(tvar: UnkindedType.Var, kindMatch: KindMatch, kenv: KindEnv): Validation[Type.Var, KindError] = tvar match {
+    case tvar@UnkindedType.Var(id, text, loc) =>
+      kenv.map.get(tvar) match {
+        // Case 1: we don't know about this kind, just ascribe it with what the context expects
+        case None => tvar.ascribedWith(kindMatch.kind).toSuccess
+        // Case 2: we know about this kind, make sure it's behaving as we expect
+        case Some(actualKind) =>
+          if (kindMatch.matches(actualKind)) {
+            Type.Var(id, actualKind, text = text).toSuccess
+          } else {
+            val expectedKind = kindMatch.kind
+            KindError.UnexpectedKind(expectedKind = expectedKind, actualKind = actualKind, loc).toFailure
+          }
+      }
+
+  }
+
+  /**
+    * Performs kinding on the given free type variable, with `kindMatch` expected from context.
+    */
+  private def visitFreeTypeVar(tvar: UnkindedType.Var, kindMatch: KindMatch): Type.Var = tvar match {
+    case UnkindedType.Var(id, text, loc) =>
+      val kind = kindMatch.kind
+      Type.Var(id, kind, text = text)
+  }
+
+  /**
+    * Performs kinding on the given type under the given kind environment, with `kindMatch` expected from context.
+    */
+  private def visitType(tpe0: UnkindedType, expectedKind: KindMatch, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[Type, KindError] = tpe0 match {
+    case tvar: UnkindedType.Var => visitTypeVar(tvar, expectedKind, kenv)
+    case UnkindedType.Cst(cst, loc) =>
+      val tycon = visitTypeConstructor(cst, root)
+      val kind = tycon.kind
+      if (expectedKind.matches(kind)) {
+        Type.Cst(tycon, loc).toSuccess
+      } else {
+        KindError.UnexpectedKind(expectedKind = expectedKind.kind, actualKind = kind, loc).toFailure
+      }
+    case UnkindedType.Apply(t10, t20) =>
```

Comment:
Whats with the t10 and t20 again :)

Reply:
is `t1_0` a better naming convention? I wanted to keep things consistent (and easy to automate) with the `0` suffix.